### ISSUE TITLE
Feature Request: Bearer auth

### ISSUE THREAD

**Issue by waptik** [honojs/hono#279]
Hey! Bearer auth (eg. `Authorization: Bearer SOME_API_KEY`) is the most commonly used type of authorization header if i'm not wrong. 
It would be great to have a bearer auth middleware like [BasicAuth](https://github.com/honojs/hono/blob/master/src/middleware/basic-auth/index.ts).

I'm currently using a hardcoded version myself, so having a built-in one will help save a lot time.

Thanks

**Comment by yusukebe** [comment]
Hi @waptik 

Thank you for your suggestion!

How about using **jwt** middleware? https://github.com/honojs/hono/tree/master/src/middleware/jwt

If you want to use `honoiscool` as a secret password, write like following

```ts
import { Hono } from 'hono'
import { jwt } from 'hono/jwt'

const app = new Hono()

app.use(
  '/auth/*',
  jwt({
    secret: 'honoiscool',
  })
)

app.get('/auth/page', (c) => {
  return c.text('You are authorized')
})

export default app
```

And generate the jwt token. It is `eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.e30.pobVINReOWuOWjdFxy2nUEjM7UQjjWbnHQSbL6KrfhY`.

https://jwt.io/#debugger-io?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.e30.pobVINReOWuOWjdFxy2nUEjM7UQjjWbnHQSbL6KrfhY

<img width="1245" alt="スクリーンショット 2022-05-27 6 57 33" src="https://user-images.githubusercontent.com/10682/170586699-b94de404-82b4-4a0c-80f8-fbdb93da2509.png">

Finaly, access end point with `Authorization: Bearer {jw token}` header.


<img width="1093" alt="スクリーンショット 2022-05-27 6 59 06" src="https://user-images.githubusercontent.com/10682/170586858-fa834eb3-a04b-406b-87b2-779ec614c555.png">

That's all!

I think it's not necessary to make another Bear auth middleware.

PS.
It's an example for using a secret token in the Env variables.

```ts
import { Hono } from 'hono'
import { jwt } from 'hono/jwt'

interface Env {
  SECRET: string
}

const app = new Hono<Env>()

app.use('/auth/*', async (c, next) => {
  const auth = jwt({
    secret: c.env.SECRET,
  })
  return await auth(c, next)
})

app.get('/auth/page', (c) => {
  return c.text('You are authorized')
})

export default app
```

**Comment by waptik** [comment]
> Hi @waptik
> 
> Thank you for your suggestion!
> 
> How about using **jwt** middleware? [https://github.com/honojs/hono/tree/master/src/middleware/jwt](https://github.com/honojs/hono/tree/master/src/middleware/jwt?rgh-link-date=2022-05-26T22%3A07%3A55Z)
> 
> If you want to use `honoiscool` as a secret password, write like following
> 
> ```ts
> import { Hono } from 'hono'
> import { jwt } from 'hono/jwt'
> 
> const app = new Hono()
> 
> app.use(
>   '/auth/*',
>   jwt({
>     secret: 'honoiscool',
>   })
> )
> 
> app.get('/auth/page', (c) => {
>   return c.text('You are authorized')
> })
> 
> export default app
> ```
> 
> And generate the jwt token. It is `eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.e30.pobVINReOWuOWjdFxy2nUEjM7UQjjWbnHQSbL6KrfhY`.
> 
> [jwt.io/#debugger-io?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.e30.pobVINReOWuOWjdFxy2nUEjM7UQjjWbnHQSbL6KrfhY](https://jwt.io/#debugger-io?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.e30.pobVINReOWuOWjdFxy2nUEjM7UQjjWbnHQSbL6KrfhY)
> 
> <img alt="スクリーンショット 2022-05-27 6 57 33" width="1245" src="https://user-images.githubusercontent.com/10682/170586699-b94de404-82b4-4a0c-80f8-fbdb93da2509.png">
> 
> Finaly, access end point with `Authorization: Bearer {jw token}` header.
> 
> <img alt="スクリーンショット 2022-05-27 6 59 06" width="1093" src="https://user-images.githubusercontent.com/10682/170586858-fa834eb3-a04b-406b-87b2-779ec614c555.png">
> 
> That's all!
> 
> I think it's not necessary to make another Bear auth middleware.
> 
> PS. It's an example for using a secret token in the Env variables.
> 
> ```ts
> import { Hono } from 'hono'
> import { jwt } from 'hono/jwt'
> 
> interface Env {
>   SECRET: string
> }
> 
> const app = new Hono<Env>()
> 
> app.use('/auth/*', async (c, next) => {
>   const auth = jwt({
>     secret: c.env.SECRET,
>   })
>   return await auth(c, next)
> })
> 
> app.get('/auth/page', (c) => {
>   return c.text('You are authorized')
> })
> 
> export default app
> ```

Thanks for the suggestion and examples. Looking at the `jwt` middleware, it seems like the secret can only be in the form of a jwt type.
I usually use `crypto.randomBytes` and I'm sure other users also any token type other than jwt.
Since this middleware already exists and no need to create an additional one, I think it'd be best to rename the name of the middleware and add an optional flag (eg. `type`) to help the function decide whether to use `jwt.verify` to verify secret.

A sample:

```plain
jwt -> bearerAuth or something else
```
Using `bearerAuth` as new name:
```ts
// function defintion
export const bearerAuth = (options: { secret: string; jwt?: { alg?: string } }) => {
	....
	return async (ctx: Context, next: Next) => {
...
		if(options.jwt) {
			try {
      				authorized = await Jwt.verify(parts[1], options.secret, options.alg as AlgorithmTypes)
    			} catch (e) {
     				 msg = `${e}`
    			}
		} else { 
			authorized = true 
		}
...
	
	}
}

```

Usage of the newly defined auth middleware

```ts
// without jwt
app.use('/auth/*', async (c, next) => {
  const auth = bearerAuth({
    secret: c.env.SECRET, // secret is in form of plain text
  })
  return await auth(c, next)
})

// with jwt
app.use('/auth/*', async (c, next) => {
  const auth = bearerAuth({
    secret: c.env.SECRET, // secret is in form of a jwt token
    jwt:{}
  })
  return await auth(c, next)
})
```

**Comment by yusukebe** [comment]
Hi @waptik !

I understood well. I think it's better to make another "*Bearer Auth middleware*". I'll consider it. Wait a minute.

**Comment by waptik** [comment]
> Hi @waptik !
> 
> I understood well. I think it's better to make another "_Bearer Auth middleware_". I'll consider it. Wait a minute.

Thanks for the consideration

**Comment by yusukebe** [comment]
Hi @waptik !

I've implemented *Bearer Auth* middleware in this PR #299 . If you can, please check whether it is what you expect.
Usage is below:

```ts
import { Hono } from 'hono'
import { bearerAuth } from 'hono/bearer-auth'

const app = new Hono()

const token = 'honoisacool'

app.use('/auth/*', bearerAuth({ token }))

app.get('/auth/page', (c) => {
  return c.text('You are authorized')
})

app.fire()
```

**Comment by waptik** [comment]
> Hi @waptik !
> 
> I've implemented _Bearer Auth_ middleware in this PR #299 . If you can, please check whether it is what you expect. Usage is below:
> 
> ```ts
> import { Hono } from 'hono'
> import { bearerAuth } from 'hono/bearer-auth'
> 
> const app = new Hono()
> 
> const token = 'honoisacool'
> 
> app.use('/auth/*', bearerAuth({ token }))
> 
> app.get('/auth/page', (c) => {
>   return c.text('You are authorized')
> })
> 
> app.fire()
> ```

Hey @yusukebe,

Yup, that's what I expect it to be. When it's released, I'll test it in production and give you feedback
